authorization/v2/uniformResourceIdentifier.go

```diff
@@ -24,6 +26,79 @@ func NewUriFromRequest(resource resources.Resource, handler gin.HandlerFunc, ctx
 	}
 }
 
+// NewURIFromString parses the string representation of a URI into the UniformResourceIdentifier struct.
+// NewURIFromString expects the string to be of the following form, otherwise ErrInvalidURI is returned.
+// hs:<service_name>:<subsystem>:<version>:<category>:<resource_name>?<allowed_arguments>#<permission_metadata>
+func NewURIFromString(source string) (UniformResourceIdentifier, error) {
+	remainingURI, metadata, err := newURIListFromString(source, "#")
+	if err != nil {
+		return UniformResourceIdentifier{}, errors.Wrap(ErrInvalidURI, err.Error())
+	}
+
+	remainingURI, arguments, err := newURIListFromString(remainingURI, "?")
+	if err != nil {
+		return UniformResourceIdentifier{}, errors.Wrap(ErrInvalidURI, err.Error())
+	}
+
+	return UniformResourceIdentifier{
+		path:      remainingURI,
+		arguments: arguments,
+		metadata:  metadata,
+	}, nil
+}
+
+func newURIListFromString(source string, sep string) (remainingURI string, uriList map[string]string, error error) {
+	sourceSplit := strings.Split(source, sep)
+
+	if len(sourceSplit) > 2 {
+		return "", nil, errors.New("malformed uri list")
+	} else if len(sourceSplit) == 2 {
+		unescapedURIList, err := url.QueryUnescape(sourceSplit[1])
+		if err != nil {
+			return "", nil, errors.Wrap(err, "could not unmarshall uri metadata")
+		}
+
+		uriList, err := unmarshallURIList(unescapedURIList)
+		if err != nil {
+			return "", nil, errors.New("could not unmarshall uri metadata")
+		}
+
+		return sourceSplit[0], uriList, nil
+	} else {
+		return source, nil, nil
+	}
+}
+
+// MarshalJSON will convert the UniformResourceIdentifier struct into the standard string representation for URIs.
+func (uri UniformResourceIdentifier) MarshalJSON() ([]byte, error) {
+	var (
+		marshalledURI      = uri.path
+		marshalledArgs     = marshallURIMap(uri.arguments)
+		marshalledMetadata = marshallURIMap(uri.metadata)
+	)
+
+	if len(marshalledArgs) > 0 {
+		marshalledURI += "?" + url.QueryEscape(marshalledArgs)
+	}
+
+	if len(marshalledMetadata) > 0 {
+		marshalledURI += "#" + url.QueryEscape(marshalledMetadata)
+	}
+
+	return []byte(fmt.Sprintf("\"%s\"", marshalledURI)), nil
+}
+
+func (uri *UniformResourceIdentifier) UnmarshalJSON(data []byte) error {
```

Comment:
Could we have a unit test for this? A real simple one will do since we're just calling `NewURIFromString` here.